### PR TITLE
Sema: Move the availability macros cache to the ASTContext

### PR DESCRIPTION
**Pull request by xymus**
The availability macros definitions are parsed from the command line and stored in a cache. The cache was in the Parser, which would have it be computed for each file using availability attributes. Let's move it to the ASTContext instead where it can generally be computed once per invocation and used across the module, speeding up the compilation process.

Also optimize how we register the buffer holding the macro definitions. Registering all of them before starting to parse their content avoids sorting them repeatedly which caused a noticeable slowdown on high usage of availability macros.

rdar://134797088

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Review by tshortli — APPROVED** [review]

File: include/swift/AST/AvailabilitySpec.h [hunk 1]
```diff
@@ -215,6 +215,19 @@ class OtherPlatformAvailabilitySpec : public AvailabilitySpec {
...
+/// Organized as two nested \c DenseMap keyed first on the macro name then
+/// the macro version. This structure allows to peek at macro names before
+/// parsing a version tuple.
+class AvailabilityMacroMap {
+public:
+  typedef llvm::DenseMap<llvm::VersionTuple,
+                         SmallVector<AvailabilitySpec *, 4>> VersionEntry;
+
+  bool WasParsed = false;
+  llvm::DenseMap<StringRef, VersionEntry> Impl;
```
Comment (DougGregor) [inline include/swift/AST/AvailabilitySpec.h:228]: Should this be a StringMap rather than depend on the StringRefs living forever?

Comment (xymus) [reply include/swift/AST/AvailabilitySpec.h:228]: I think that's safe currently. The StringRefs should point to a buffer attached to the SourceManager of the ASTContext, so both the cache and the sources should be freed together. Does this sound right?

Comment (xymus) [reply include/swift/AST/AvailabilitySpec.h:228]: Currently one of these buffers is allocated for each `-define-availability` argument. I suspect this doesn't help with the general performances of the `findBuffer` services on wide uses of availability macros, we could look into deregistering these small buffers after parsing and copy all these strings.

File: lib/AST/ASTContext.cpp [hunk 2]
```diff
@@ -426,6 +426,9 @@ struct ASTContext::Implementation {
   /// Singleton used to cache the import graph.
   swift::namelookup::ImportCache TheImportCache;
 
+  /// Cache of availability macros parsed from the command line.
+  AvailabilityMacroMap TheAvailabilityMacroCache;
```
Comment (DougGregor) [inline lib/AST/ASTContext.cpp:430]: Could we use the request-evaluator here rather than add the cache directly to the ASTContext?

Comment (xymus) [reply lib/AST/ASTContext.cpp:430]: Other requests seem to use the ASTContext itself. I'll see to use that.

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
Merging with 2 different optimizations. I'll put up a PR to move the cache to a request next as this optimization is time sensitive.

**Pull request merged**